gdpr/validate_block.go

```diff
@@ -0,0 +1,247 @@
+					fmt.Printf("that's odd")
+				}
+				var b = getHash(kvWrite.GetValue())
+				//kvWrite.ValueHash = []byte("Gal")
+				if memberOf((string)(b),m) == false{
+					return nil,  ErrVal
+				}
+
+			}
+
+		}
+	}
+
+	return getVanillaBlock(block)
+
+	//return nil, nil
+}
+
+
+
+
+func memberOf(a string, m map[string]struct{}) bool {
+	_ , exists := m[a]
+	return exists
+
+}
+
+func getVanillaBlock(block *common.Block) (*common.Block , error) {
+	newBlock := proto.Clone(block).(*common.Block)
+	newBlock.Data.PreimageSpace = nil
+
+	clearKVWrites(newBlock) // TODO: remove when blocks are generated without KVWrite values.
+	//preImages := block.GetData().GetPreimageSpace() // TODO: to be used in real world
+	preImages := extractPreimages(block)
+
+	for _, envBytes := range newBlock.Data.Data {
+		env, err := protoutil.GetEnvelopeFromBlock(envBytes)
+		if err != nil {
+			//logger.Warning("Invalid envelope:", err)
+			return nil,err
+		}
+
+		payload, err := protoutil.UnmarshalPayload(env.Payload) //protoutil.GetBytesPayload()
+		if err != nil {
+			//logger.Warning("Invalid payload:", err)
+			return nil,err
+		}
+
+		tx, err := protoutil.UnmarshalTransaction(payload.Data)
+		if err != nil {
+			return nil,err
+		}
+
+		_, respPayload, err := protoutil.GetPayloads(tx.Actions[0])
+		if err != nil {
+			return nil,err
+		}
+
+		txRWSet := &rwsetutil.TxRwSet{}
+
+		if err = txRWSet.FromProtoBytes(respPayload.Results); err != nil {
+			return nil,err
+		}
+		for _, nsRWSet := range txRWSet.NsRwSets {
+			for _, kvWrite := range nsRWSet.KvRwSet.Writes {
+				var a = kvWrite.GetValueHash()
+				//TODO: kvWrite.ValueHash := nil ??
+				temp := getKVWrite(preImages, a)
+				if temp != nil {
+					kvWrite.Value = temp
+				}
+			}
+
+		}
+	}
+
+	return newBlock,nil
+
+
+}
+
+func getKVWrite(preimages [][]byte , hashVal []byte ) []byte {
+	strVal := (string)(hashVal)
+	for _, pi := range preimages {
+		temp := (string)(getHash(pi))
+		if temp == strVal{
+			return pi
+		}
+	}
+	return nil
+}
+
+
+func extractPreimages(block *common.Block) [][]byte {
+
+	var preimages [][]byte
+
+	for _, envBytes := range block.Data.Data {
+		env, err := protoutil.GetEnvelopeFromBlock(envBytes)
+		if err != nil {
+			//logger.Warning("Invalid envelope:", err)
+			continue
+		}
+
+		payload, err := protoutil.UnmarshalPayload(env.Payload) //protoutil.GetBytesPayload()
+		if err != nil {
+			//logger.Warning("Invalid payload:", err)
+			continue
+		}
+
+		tx, err := protoutil.UnmarshalTransaction(payload.Data)
+		if err != nil {
+			return nil
+		}
+
+		_, respPayload, err := protoutil.GetPayloads(tx.Actions[0])
+		if err != nil {
+			return nil
+		}
+
+		txRWSet := &rwsetutil.TxRwSet{}
+
+		if err = txRWSet.FromProtoBytes(respPayload.Results); err != nil {
+			return nil
+		}
+		for _, nsRWSet := range txRWSet.NsRwSets {
+			//nsRWSet.KvRwSet
+			for _, kvWrite := range nsRWSet.KvRwSet.Writes {
+				var a = kvWrite.GetValue()
+				preimages = append(preimages, a)
```

Comment:
the pre-images should not appear here, they belong to the pre-image space. if you want to extract them from a block you need to assume they are in the right place.